[ao] Added filtered table generation capability to ModelReportVisualizer

vspenubarthi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81975
* #81974
* #81973
* __->__ #81673

Summary: This adds the ability to generate and display the collected
statistics in a table format for the ModelReportVisualizer. The output
of this is a dictionary containing two keys, mapping to a tensor stats
table and channel stats table respectively.

The two ways you can filter is by module_fqn, by only including modules
with the `module_fqn_filter` substring, or by feature filter, which only includes
features that contain the `feature_filter` substring.

Expected Use:
```
table_dict = mod_rep_visualizer.generate_filtered_tables()

tensor_table = table_dict[ModelReportVisualizer.TABLE_TENSOR_KEY]
channel_table = table_dict[ModelReportVisualizer.TABLE_CHANNEL_KEY]
```

Headers for the Tensor level info:
```
         idx  layer_fqn  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  ---------   ---------   ---------        ---------
```

Headers for the channel level info:
```
         idx  layer_fqn  channel  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  -------  ---------   ---------   ---------        ---------
```

The reason we split this up into two tables is because with the design
where everything is in one table, it is ambiguous and easy to mix up
whether a tensor level stat is actually tensor level stat or might be a
per channel stat since we would have a row for each channel.

Also changed some of the framework to abstract out the finding of the
tables to the actual visualization to make the API much easier for the
user to digest and parse.

Test Plan: python test/test_quantization.py TestFxModelReportVisualizer.test_generate_table

Reviewers:

Subscribers:

Tasks:

Tags:

vspenubarthi committed:
[ao][work-in-progress] Added table generation capability to ModelReportVisualizer

Summary: This adds the ability to generate and display the collected
statistics in a table format for the ModelReportVisualizer. The output
of this is a tuple containing a dictionary mapping to tensor level info
and to channel level info, and a formatted string that can be printed to
view a table.

The two ways you can filter is by module_fqn, by only including modules
with the `module_fqn_filter` substring, or by feature filter, which only includes
features that contain the `feature_filter` substring.

Expected Use:
```
info, tabluated_str = model_report_visualizer.generate_table_view(*filters)
print(tabulated_str) # outputs neatly formatted table
```

Headers for the Tensor level info:
```
         idx  layer_fqn  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  ---------   ---------   ---------        ---------
```

Headers for the channel level info:
```
         idx  layer_fqn  channel  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  -------  ---------   ---------   ---------        ---------
```

The reason we split this up into two tables is because with the design
where everything is in one table, it is ambiguous and easy to mix up
whether a tensor level stat is actually tensor level stat or might be a
per channel stat since we would have a row for each channel.

Still need to flesh out the test case.

Test Plan: python test/test_quantization.py TestFxModelReportVisualizer.test_generate_table

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao][work-in-progress] Added table generation capability to ModelReportVisualizer"


Summary: This adds the ability to generate and display the collected
statistics in a table format for the ModelReportVisualizer. The output
of this is a tuple containing a dictionary mapping to tensor level info
and to channel level info, and a formatted string that can be printed to
view a table.

The two ways you can filter is by module_fqn, by only including modules
with the `module_fqn_filter` substring, or by feature filter, which only includes
features that contain the `feature_filter` substring.

Expected Use:
```
info, tabluated_str = model_report_visualizer.generate_table_view(*filters)
print(tabulated_str) # outputs neatly formatted table
```

Headers for the Tensor level info:
```
         idx  layer_fqn  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  ---------   ---------   ---------        ---------
```

Headers for the channel level info:
```
         idx  layer_fqn  channel  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  -------  ---------   ---------   ---------        ---------
```

The reason we split this up into two tables is because with the design
where everything is in one table, it is ambiguous and easy to mix up
whether a tensor level stat is actually tensor level stat or might be a
per channel stat since we would have a row for each channel.

Still need to flesh out the test case.

Test Plan: python test/test_quantization.py TestFxModelReportVisualizer.test_generate_table

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao][work-in-progress] Added table generation capability to ModelReportVisualizer"


Summary: This adds the ability to generate and display the collected
statistics in a table format for the ModelReportVisualizer. The output
of this is a tuple containing a dictionary mapping to tensor level info
and to channel level info, and a formatted string that can be printed to
view a table.

The two ways you can filter is by module_fqn, by only including modules
with the `module_fqn_filter` substring, or by feature filter, which only includes
features that contain the `feature_filter` substring.

Expected Use:
```
info, tabluated_str = model_report_visualizer.generate_table_view(*filters)
print(tabulated_str) # outputs neatly formatted table
```

Headers for the Tensor level info:
```
         idx  layer_fqn  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  ---------   ---------   ---------        ---------
```

Headers for the channel level info:
```
         idx  layer_fqn  channel  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  -------  ---------   ---------   ---------        ---------
```

The reason we split this up into two tables is because with the design
where everything is in one table, it is ambiguous and easy to mix up
whether a tensor level stat is actually tensor level stat or might be a
per channel stat since we would have a row for each channel.

Still need to flesh out the test case.

Test Plan: python test/test_quantization.py TestFxModelReportVisualizer.test_generate_table

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added filtered table generation capability to ModelReportVisualizer"


Summary: This adds the ability to generate and display the collected
statistics in a table format for the ModelReportVisualizer. The output
of this is a dictionary containing two keys, mapping to a tensor stats
table and channel stats table respectively.

The two ways you can filter is by module_fqn, by only including modules
with the `module_fqn_filter` substring, or by feature filter, which only includes
features that contain the `feature_filter` substring.

Expected Use:
```
table_dict = mod_rep_visualizer.generate_filtered_tables()

tensor_table = table_dict[ModelReportVisualizer.TABLE_TENSOR_KEY]
channel_table = table_dict[ModelReportVisualizer.TABLE_CHANNEL_KEY]
```

Headers for the Tensor level info:
```
         idx  layer_fqn  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  ---------   ---------   ---------        ---------
```

Headers for the channel level info:
```
         idx  layer_fqn  channel  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  -------  ---------   ---------   ---------        ---------
```

The reason we split this up into two tables is because with the design
where everything is in one table, it is ambiguous and easy to mix up
whether a tensor level stat is actually tensor level stat or might be a
per channel stat since we would have a row for each channel.

Also changed some of the framework to abstract out the finding of the
tables to the actual visualization to make the API much easier for the
user to digest and parse.

Test Plan: python test/test_quantization.py TestFxModelReportVisualizer.test_generate_table

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -1715,6 +1715,77 @@ def test_get_modules_and_features(self):
+            detector_set.add(OutlierDetector(reference_percentile=0.95))
+            detector_set.add(InputWeightEqualizationDetector(0.5))
+
+            model = TestFxModelReportClass.TwoThreeOps()
```
nit: maybe just put the TwoThreeOps class in the top level of this file

vspenubarthi replied:
done

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -1715,6 +1715,77 @@ def test_get_modules_and_features(self):
+            table_dict = mod_rep_visualizer.generate_filtered_tables()
+
+            # test primarily the dict since it has same info as str
+            tensor_table = table_dict[ModelReportVisualizer.TABLE_TENSOR_KEY]
```
user still need to know the structure of the dictionary in this case, should we make this a dataclass object or something so that user can call some methods to get it?

vspenubarthi replied:
i feel like for now, the dictionary has relatively few keys, so they should be able to just easily look at the keys (and I don't expect it to grow anymore).

However, I think revisiting these and making them data classes is an interesting idea worth revisiting later.

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -1715,6 +1715,77 @@ def test_get_modules_and_features(self):
+            tensor_info_modules = set(row[1] for row in tensor_table[1:])
+            channel_info_modules = set(row[1] for row in channel_table[1:])
+            combined_modules: Set = tensor_info_modules.union(channel_info_modules)
+            self.assertEqual(len(combined_modules), 0)  # should be no matching modules
```
I feel ideally this should be separate test

vspenubarthi replied:
i think the reason i kept this in here even though it is testing more of the number of modules is because I wanted to organize my tests by the primary methods i was testing, so i thought it made sense to keep it in here. 

If you strongly feel like it should be moved I can move it. Should this made a separate test combined with the other one below in this case?

vspenubarthi replied:
**made it its own test in the end, made more sense, thanks for the feedback**

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -1715,6 +1715,77 @@ def test_get_modules_and_features(self):
+            channel_info_features = len(channel_table[0]) - ModelReportVisualizer.DEFAULT_NON_FEATURE_CHANNEL_HEADERS
+
+            # make sure that there are no tensor features, and that there is one channel level feature
+            self.assertEqual(tensor_info_features, 0)
+            self.assertEqual(channel_info_features, 1)
```
same for this one

vspenubarthi replied:
i think the reason i kept this in here even though it is testing more of the features is because I wanted to organize my tests by the primary methods i was testing, so i thought it made sense to keep it in here. 

If you strongly feel like it should be moved I can move it. Should this made a separate test combined with the other one above in this case?

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -102,135 +112,256 @@ def get_all_unique_feature_names(self, plottable_features_only: bool = True) ->
+                contain this filter substring
                 Default = "", results in all the features being printed
-            module_fqn_prefix_filter (str, optional): Only includes modules with this string prefix
+            module_fqn_filter (str, optional): Only includes modules with this string
```
you mean only includes modules with names that contains this string?

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -102,135 +112,256 @@ def get_all_unique_feature_names(self, plottable_features_only: bool = True) ->
+                feature_data = filtered_data[module_fqn][feature_name]
+
+                # if it is only a single value, is tensor, otherwise per channel
+                try:
+                    iterator = iter(feature_data)
+                except TypeError:
```
is there an easier way to check? e.g. through types? `using try/catch to check something is not recommended`

vspenubarthi replied:
Yeah I looked up a bunch of things on stackoverflow and other sites and the recommended method everybody said to use for Python specifically is try catch, which people are calling the "pythonic" way of doing things even though it's not really popular with other languages.

After looking up multiple resources, it doesn't look like there is a more elegant way for now. I tried making it checking for types specifically but i got errors like 

```
    raise TypeError("len() of a 0-d tensor")
TypeError: len() of a 0-d tensor
```
in my tests because you can have a single value tensor.

jerryzh168 replied:
how are you checking for types?
does `isinstance(feature_data, list)` work?

vspenubarthi replied:
this is what i tested but because some of the values are lists, and others are tensors, this alone doesn't work. And I also tried `isinstance(feature_data, torch.Tensor)` but that didn't work either because of single value tensors. Updated to have if else statements instead.

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -102,135 +112,256 @@ def get_all_unique_feature_names(self, plottable_features_only: bool = True) ->
+                        feature_val = "Not Applicable"
+
+                    # if it's a tensor we want to extract val
+                    if type(feature_val) is torch.Tensor:
```
nit: maybe `isinstance(feature_val, torch.Tensor)`

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -102,135 +112,256 @@ def get_all_unique_feature_names(self, plottable_features_only: bool = True) ->
+        if len(channel_table) == 0:
+            channel_table.append([])  # just an empty list to signify no data
+        else:
+            channel_headers: List[str] = ["idx", "layer_fqn", "channel"] + channel_features_list
```
should headers be a part of the table as well? this looks like a metadata for the table. maybe we should define a object to represent table?

vspenubarthi replied:
i don't feel like the table is enough of a thing that it warrants its own class, it's only attributes are its headers and the table, however if you believe it should get one we can make the change

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -102,135 +112,256 @@ def get_all_unique_feature_names(self, plottable_features_only: bool = True) ->
+        return filtered_dict
 
-    def generate_table_visualization(self, feature: str = "", module_fqn_prefix_filter: str = ""):
+    def generate_filtered_tables(self, feature_filter: str = "", module_fqn_filter: str = "") -> Dict[str, List[List[Any]]]:
```
this method is pretty long, can we split it into multiple smaller functions?

vspenubarthi replied:
broke this up by adding two helper functions, each one is much smaller now

vspenubarthi committed:
Update on "[ao] Added filtered table generation capability to ModelReportVisualizer"


Summary: This adds the ability to generate and display the collected
statistics in a table format for the ModelReportVisualizer. The output
of this is a dictionary containing two keys, mapping to a tensor stats
table and channel stats table respectively.

The two ways you can filter is by module_fqn, by only including modules
with the `module_fqn_filter` substring, or by feature filter, which only includes
features that contain the `feature_filter` substring.

Expected Use:
```
table_dict = mod_rep_visualizer.generate_filtered_tables()

tensor_table = table_dict[ModelReportVisualizer.TABLE_TENSOR_KEY]
channel_table = table_dict[ModelReportVisualizer.TABLE_CHANNEL_KEY]
```

Headers for the Tensor level info:
```
         idx  layer_fqn  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  ---------   ---------   ---------        ---------
```

Headers for the channel level info:
```
         idx  layer_fqn  channel  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  -------  ---------   ---------   ---------        ---------
```

The reason we split this up into two tables is because with the design
where everything is in one table, it is ambiguous and easy to mix up
whether a tensor level stat is actually tensor level stat or might be a
per channel stat since we would have a row for each channel.

Also changed some of the framework to abstract out the finding of the
tables to the actual visualization to make the API much easier for the
user to digest and parse.

Test Plan: python test/test_quantization.py TestFxModelReportVisualizer.test_generate_table

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added filtered table generation capability to ModelReportVisualizer"


Summary: This adds the ability to generate and display the collected
statistics in a table format for the ModelReportVisualizer. The output
of this is a dictionary containing two keys, mapping to a tensor stats
table and channel stats table respectively.

The two ways you can filter is by module_fqn, by only including modules
with the `module_fqn_filter` substring, or by feature filter, which only includes
features that contain the `feature_filter` substring.

Expected Use:
```
table_dict = mod_rep_visualizer.generate_filtered_tables()

tensor_table = table_dict[ModelReportVisualizer.TABLE_TENSOR_KEY]
channel_table = table_dict[ModelReportVisualizer.TABLE_CHANNEL_KEY]
```

Headers for the Tensor level info:
```
         idx  layer_fqn  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  ---------   ---------   ---------        ---------
```

Headers for the channel level info:
```
         idx  layer_fqn  channel  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  -------  ---------   ---------   ---------        ---------
```

The reason we split this up into two tables is because with the design
where everything is in one table, it is ambiguous and easy to mix up
whether a tensor level stat is actually tensor level stat or might be a
per channel stat since we would have a row for each channel.

Also changed some of the framework to abstract out the finding of the
tables to the actual visualization to make the API much easier for the
user to digest and parse.

Test Plan: python test/test_quantization.py TestFxModelReportVisualizer.test_generate_table

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added filtered table generation capability to ModelReportVisualizer"


Summary: This adds the ability to generate and display the collected
statistics in a table format for the ModelReportVisualizer. The output
of this is a dictionary containing two keys, mapping to a tensor stats
table and channel stats table respectively.

The two ways you can filter is by module_fqn, by only including modules
with the `module_fqn_filter` substring, or by feature filter, which only includes
features that contain the `feature_filter` substring.

Expected Use:
```
table_dict = mod_rep_visualizer.generate_filtered_tables()

tensor_table = table_dict[ModelReportVisualizer.TABLE_TENSOR_KEY]
channel_table = table_dict[ModelReportVisualizer.TABLE_CHANNEL_KEY]
```

Headers for the Tensor level info:
```
         idx  layer_fqn  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  ---------   ---------   ---------        ---------
```

Headers for the channel level info:
```
         idx  layer_fqn  channel  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  -------  ---------   ---------   ---------        ---------
```

The reason we split this up into two tables is because with the design
where everything is in one table, it is ambiguous and easy to mix up
whether a tensor level stat is actually tensor level stat or might be a
per channel stat since we would have a row for each channel.

Also changed some of the framework to abstract out the finding of the
tables to the actual visualization to make the API much easier for the
user to digest and parse.

Test Plan: python test/test_quantization.py TestFxModelReportVisualizer.test_generate_table

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added filtered table generation capability to ModelReportVisualizer"


Summary: This adds the ability to generate and display the collected
statistics in a table format for the ModelReportVisualizer. The output
of this is a dictionary containing two keys, mapping to a tensor stats
table and channel stats table respectively.

The two ways you can filter is by module_fqn, by only including modules
with the `module_fqn_filter` substring, or by feature filter, which only includes
features that contain the `feature_filter` substring.

Expected Use:
```
table_dict = mod_rep_visualizer.generate_filtered_tables()

tensor_table = table_dict[ModelReportVisualizer.TABLE_TENSOR_KEY]
channel_table = table_dict[ModelReportVisualizer.TABLE_CHANNEL_KEY]
```

Headers for the Tensor level info:
```
         idx  layer_fqn  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  ---------   ---------   ---------        ---------
```

Headers for the channel level info:
```
         idx  layer_fqn  channel  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  -------  ---------   ---------   ---------        ---------
```

The reason we split this up into two tables is because with the design
where everything is in one table, it is ambiguous and easy to mix up
whether a tensor level stat is actually tensor level stat or might be a
per channel stat since we would have a row for each channel.

Also changed some of the framework to abstract out the finding of the
tables to the actual visualization to make the API much easier for the
user to digest and parse.

Test Plan: python test/test_quantization.py TestFxModelReportVisualizer.test_generate_table

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -102,135 +112,305 @@ def get_all_unique_feature_names(self, plottable_features_only: bool = True) ->
-            feature (str): The specific feature we wish to generate the plot for
-            module_fqn_prefix_filter (str, optional): Only includes modules with this string prefix
-                Default = "", results in all the modules in the reports to be visible in the histogram
+            feature_filter (str, optional): Filters the features presented to only those that
```
can we give some example for this? not sure what feature means

vspenubarthi replied:
added examples for all of them

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -102,135 +112,305 @@ def get_all_unique_feature_names(self, plottable_features_only: bool = True) ->
-        Expected Use:
-            >>> model_report_visualizer.generate_table_visualization(*filters)  # outputs neatly formatted table
+        Returns a dictionary with two keys:
+            (Dict[strList[List[Any]]]) A dict containing two keys:
```
`Dict[str, List[List[Any]]]`?

vspenubarthi committed:
Update on "[ao] Added filtered table generation capability to ModelReportVisualizer"


Summary: This adds the ability to generate and display the collected
statistics in a table format for the ModelReportVisualizer. The output
of this is a dictionary containing two keys, mapping to a tensor stats
table and channel stats table respectively.

The two ways you can filter is by module_fqn, by only including modules
with the `module_fqn_filter` substring, or by feature filter, which only includes
features that contain the `feature_filter` substring.

Expected Use:
```
table_dict = mod_rep_visualizer.generate_filtered_tables()

tensor_table = table_dict[ModelReportVisualizer.TABLE_TENSOR_KEY]
channel_table = table_dict[ModelReportVisualizer.TABLE_CHANNEL_KEY]
```

Headers for the Tensor level info:
```
         idx  layer_fqn  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  ---------   ---------   ---------        ---------
```

Headers for the channel level info:
```
         idx  layer_fqn  channel  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  -------  ---------   ---------   ---------        ---------
```

The reason we split this up into two tables is because with the design
where everything is in one table, it is ambiguous and easy to mix up
whether a tensor level stat is actually tensor level stat or might be a
per channel stat since we would have a row for each channel.

Also changed some of the framework to abstract out the finding of the
tables to the actual visualization to make the API much easier for the
user to digest and parse.

Test Plan: python test/test_quantization.py TestFxModelReportVisualizer.test_generate_table

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -102,138 +112,337 @@ def get_all_unique_feature_names(self, plottable_features_only: bool = True) ->
+        Currently meant to generate the headers and table for both the tensor information.
+
+        Args:
+            filtered_data (OrderedDict[str, Any]): An OrderedDict (sorted in order of model) mapping:
```
should this be OrderedDict[str, Dict[str, Any]]?

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -102,138 +112,337 @@ def get_all_unique_feature_names(self, plottable_features_only: bool = True) ->
+                feature_data = filtered_data[module_fqn][feature_name]
+
+                # if it is only a single value, is tensor, otherwise per channel
+                if isinstance(feature_data, torch.Tensor):
+                    # see what the size of shape is
+                    if len(feature_data.shape) == 0:
+                        # single value
+                        # means is per-tensor
+                        tensor_features.add(feature_name)
+                    else:
+                        # works means per channel
+                        channel_features.add(feature_name)
+                        num_channels = len(feature_data)
+                elif isinstance(feature_data, list):
+                    # works means per channel
+                    channel_features.add(feature_name)
+                    num_channels = len(feature_data)
+                else:
+                    # means is per-tensor
+                    tensor_features.add(feature_name)
+
```
can we merge the per tensor cases and per channel cases into the same branch so it's clearer?

vspenubarthi replied:
merged

jerryzh168 reviewed (approved):
lgtm, had some comments inline

vspenubarthi committed:
Update on "[ao] Added filtered table generation capability to ModelReportVisualizer"


Summary: This adds the ability to generate and display the collected
statistics in a table format for the ModelReportVisualizer. The output
of this is a dictionary containing two keys, mapping to a tensor stats
table and channel stats table respectively.

The two ways you can filter is by module_fqn, by only including modules
with the `module_fqn_filter` substring, or by feature filter, which only includes
features that contain the `feature_filter` substring.

Expected Use:
```
table_dict = mod_rep_visualizer.generate_filtered_tables()

tensor_table = table_dict[ModelReportVisualizer.TABLE_TENSOR_KEY]
channel_table = table_dict[ModelReportVisualizer.TABLE_CHANNEL_KEY]
```

Headers for the Tensor level info:
```
         idx  layer_fqn  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  ---------   ---------   ---------        ---------
```

Headers for the channel level info:
```
         idx  layer_fqn  channel  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  -------  ---------   ---------   ---------        ---------
```

The reason we split this up into two tables is because with the design
where everything is in one table, it is ambiguous and easy to mix up
whether a tensor level stat is actually tensor level stat or might be a
per channel stat since we would have a row for each channel.

Also changed some of the framework to abstract out the finding of the
tables to the actual visualization to make the API much easier for the
user to digest and parse.

Test Plan: python test/test_quantization.py TestFxModelReportVisualizer.test_generate_table

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi commented:
@pytorchbot merge -g

vspenubarthi commented:
@pytorchbot merge

vspenubarthi committed:
Update on "[ao] Added filtered table generation capability to ModelReportVisualizer"


Summary: This adds the ability to generate and display the collected
statistics in a table format for the ModelReportVisualizer. The output
of this is a dictionary containing two keys, mapping to a tensor stats
table and channel stats table respectively.

The two ways you can filter is by module_fqn, by only including modules
with the `module_fqn_filter` substring, or by feature filter, which only includes
features that contain the `feature_filter` substring.

Expected Use:
```
table_dict = mod_rep_visualizer.generate_filtered_tables()

tensor_table = table_dict[ModelReportVisualizer.TABLE_TENSOR_KEY]
channel_table = table_dict[ModelReportVisualizer.TABLE_CHANNEL_KEY]
```

Headers for the Tensor level info:
```
         idx  layer_fqn  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  ---------   ---------   ---------        ---------
```

Headers for the channel level info:
```
         idx  layer_fqn  channel  feature_1   feature_2   feature_3   .... feature_n
        ----  ---------  -------  ---------   ---------   ---------        ---------
```

The reason we split this up into two tables is because with the design
where everything is in one table, it is ambiguous and easy to mix up
whether a tensor level stat is actually tensor level stat or might be a
per channel stat since we would have a row for each channel.

Also changed some of the framework to abstract out the finding of the
tables to the actual visualization to make the API much easier for the
user to digest and parse.

Test Plan: python test/test_quantization.py TestFxModelReportVisualizer.test_generate_table

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi commented:
@pytorchbot merge -g
